src/rsc.rs

```diff
@@ -0,0 +1,192 @@
+use std::sync::atomic::{AtomicBool, Ordering};
+use rpki::uri;
+use rpki::repository::rsc;
+use rpki::repository::cert::ResourceCert;
+use rpki::repository::error::{ValidationError, VerificationError};
+use rpki::repository::rsc::{ResourceSignedChecklist, Rsc};
+use rpki::repository::tal::{Tal, TalUri};
+use crate::config::Config;
+use crate::engine::{CaCert, ProcessPubPoint, ProcessRun, Engine};
+use crate::error::{Failed, RunFailed};
+
+
+//------------ ValidationReport ----------------------------------------------
+
+/// The result of an RSC validation run.
+#[derive(Debug)]
+pub struct ValidationReport {
+    rsc: rsc::Rsc,
+    complete: AtomicBool,
+    strict: bool,
+}
+
+impl ValidationReport {
+    pub fn new(
+        rsc: Rsc, config: &Config
+    ) -> Result<Self, ValidationError> {
+        let cert = rsc.signed().cert();
+        rsc.as_resources().verify_covered(cert.as_resources())
+            .map_err(|err| VerificationError::new(err.to_string()))?;
+        rsc.v4_resources().verify_covered(cert.v4_resources())
+            .map_err(|err| VerificationError::new(err.v4().to_string()))?;
+        rsc.v6_resources().verify_covered(cert.v6_resources())
+            .map_err(|err| VerificationError::new(err.v6().to_string()))?;
+        Ok(Self {
+            rsc,
+            complete: AtomicBool::new(false),
+            strict: config.strict
+        })
+    }
+
+    pub fn process(
+        &self,
+        engine: &Engine,
+    ) -> Result<(), RunFailed> {
+        let mut run = engine.start(self, false)?;
+        run.process()?;
+        run.cleanup()?;
+        Ok(())
+    }
+
+    pub fn finalize(self) -> Result<ResourceSignedChecklist, Failed> {
+        if self.complete.load(Ordering::Relaxed) {
+            Ok(self.rsc.content().clone())
+        } else {
+            Err(Failed)
+        }
+    }
+}
+
+impl ValidationReport {
+    fn supply_tal(&self, tal: &Tal) -> Result<bool, ValidationError> {
+        if self.complete.load(Ordering::Relaxed) {
+            return Ok(true);
+        }
+
+        let cert = self.rsc.signed().cert();
+
+        if cert.is_self_signed() {
```

Comment:
This would suggest that the EE cert of the RSC is the TA certificate. I don’t think that can happen?